File: server/clients/tmdb_client.py
"""
TMDB client — parses free-text queries and resolves them to structured media info.

Query parsing examples
----------------------
"Inception"                  → movie search
"Breaking Bad"               → tv search
"Breaking Bad S01"           → tv, season=1
"Breaking Bad S01E03"        → tv, season=1, episode=3
"Breaking Bad season 2"      → tv, season=2
"Attack on Titan"            → anime (auto-detected via TMDB genres)
"https://www.imdb.com/title/tt1630029/"  → direct IMDb lookup
"""
from __future__ import annotations

import re
import logging
from dataclasses import dataclass, field
from typing import Optional

from .base_client import BaseAPIClient

logger = logging.getLogger(__name__)

TMDB_BASE = "https://api.themoviedb.org/3/"

# Anime detection: TMDB genre id 16 = Animation; we also check origin_country.
ANIME_KEYWORDS = re.compile(
    r"\b(anime|manga|ova|ona|shounen|shonen|shoujo|seinen|isekai|mecha)\b", re.I
)

# Season/episode patterns
_SE_FULL = re.compile(r"[Ss](\d{1,2})[Ee](\d{1,3})")   # S01E03
_S_ONLY = re.compile(r"[Ss](\d{1,2})\b")                # S01
_SEASON_WORD = re.compile(r"\bseason\s+(\d{1,2})\b", re.I)
_EPISODE_WORD = re.compile(r"\bepisode\s+(\d{1,3})\b", re.I)
_IMDB_URL = re.compile(r"imdb\.com/title/(tt\d+)")
_YEAR_TRAIL = re.compile(r"\s*\(?(19|20)\d{2}\)?$")


@dataclass
class MediaInfo:
    title: str
    year: Optional[int] = None
    imdb_id: Optional[str] = None
    tmdb_id: Optional[int] = None
    type: str = "movie"          # "movie" | "tv" | "anime"
    season: Optional[int] = None
    episode: Optional[int] = None
    overview: Optional[str] = None
    poster_path: Optional[str] = None   # e.g. "/abc123.jpg" — combine with TMDB image base
    # TV details
    total_seasons: Optional[int] = None
    episodes_in_season: Optional[int] = None   # populated if season is set
    episode_titles: dict[int, str] = field(default_factory=dict)  # {ep_num: title}
    # Anime flag separate from type so we can still distinguish
    is_anime: bool = False

    @property
    def poster_url(self) -> Optional[str]:
        """Full TMDB poster URL at w500 size, or None."""
        if self.poster_path:
            return f"https://image.tmdb.org/t/p/w500{self.poster_path}"
        return None

    @property
    def display_name(self) -> str:
        base = f"{self.title} ({self.year})" if self.year else self.title
        if self.season and self.episode:
            return f"{base} S{self.season:02d}E{self.episode:02d}"
        if self.season:
            return f"{base} Season {self.season}"
        return base


class TMDBClient(BaseAPIClient):
    def __init__(self, api_key: str):
        super().__init__(
            base_url=TMDB_BASE,
            timeout=15,
            max_retries=3,
            backoff_base=1.0,
            params={"api_key": api_key},
        )
        self._key = api_key

    async def _get(self, path: str, **kwargs) -> dict:
        """GET a TMDB endpoint and return parsed JSON.

        Strips leading "/" so httpx resolves relative to base_url
        (a leading "/" would discard the /3/ path component per RFC 3986).
        """
        path = path.lstrip("/")
        r = await self.get(path, **kwargs)
        return r.json()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    async def search(self, raw_query: str) -> MediaInfo:
        """Parse *raw_query* and resolve it to a fully-populated MediaInfo."""
        query, season, episode = self._parse_query(raw_query.strip())

        # Direct IMDb URL?
        imdb_match = _IMDB_URL.search(raw_query)
        if imdb_match:
            return await self._lookup_imdb(imdb_match.group(1), season, episode)

        # Try movie first if no season/episode hints; otherwise favour TV
        if season is not None or episode is not None:
            info = await self._search_tv(query, season, episode)
        else:
            # Try multi-search and pick highest-confidence result
            info = await self._search_multi(query, season, episode)

        return info

    async def get_episode_count(self, tmdb_id: int, season: int) -> int:
        """Return the number of episodes in a given TV season."""
        try:
            data = await self._get(f"tv/{tmdb_id}/season/{season}")
            return len(data.get("episodes", []))
        except Exception as exc:
            logger.warning("get_episode_count failed (tmdb=%d, s=%d): %s", tmdb_id, season, exc)
            return 0

    async def get_episode_title(self, tmdb_id: int, season: int, episode: int) -> str:
        try:
            data = await self._get(f"tv/{tmdb_id}/season/{season}/episode/{episode}")
            return data.get("name", "")
        except Exception as exc:
            logger.warning("get_episode_title failed (tmdb=%d, s=%d, e=%d): %s", tmdb_id, season, episode, exc)
            return ""

    async def fuzzy_resolve(
        self,
        title: str,
        media_type: str = "movie",
        year: Optional[int] = None,
    ) -> tuple[str, Optional[int], Optional[str]]:
        """Find the best TMDB match for a title using progressive fallback.

        Returns ``(canonical_title, canonical_year, poster_path)`` where
        ``poster_path`` is the raw TMDB path (e.g. ``"/abc123.jpg"``).

        Search strategy (stops at the first pass that returns results):
          1. Type-specific endpoint  + year filter
          2. Type-specific endpoint  (no year)
          3. ``/search/multi``       with full title   (TMDB's own fuzzy engine)
          4. ``/search/multi``       progressively shorter title (strips trailing
             words one at a time, up to 4 removals, stops at 2-word minimum)

        Results are scored by (exact-title-and-year, exact-title, popularity)
        so a precise match always wins even if found during a fuzzy pass.
        """
        is_tv     = media_type in ("tv", "anime")
        endpoint  = "search/tv" if is_tv else "search/movie"
        date_key  = "first_air_date_year" if is_tv else "year"
        title_key = "name" if is_tv else "title"

        def _score(item: dict) -> tuple:
            t  = (item.get("title") or item.get("name") or "").lower()
            ds = item.get("release_date") or item.get("first_air_date") or ""
            iy = int(ds[:4]) if ds[:4].isdigit() else 0
            return (
                t == title.lower() and (iy == year if year else True),
                t == title.lower(),
                item.get("popularity", 0),
            )

        def _extract(best: dict, tkey: str) -> tuple[str, Optional[int], Optional[str]]:
            ds = best.get("release_date") or best.get("first_air_date") or ""
            return (
                best.get(tkey) or title,
                int(ds[:4]) if ds[:4].isdigit() else year,
                best.get("poster_path"),
            )

        async def _typed(q: str, with_year: bool) -> list[dict]:
            p: dict = {"query": q, "include_adult": False}
            if with_year and year:
                p[date_key] = year
            data = await self._get(endpoint, params=p)
            return data.get("results", [])

        async def _multi(q: str) -> list[dict]:
            data = await self._get(
                "search/multi", params={"query": q, "include_adult": False}
            )
            return [
                x for x in data.get("results", [])
                if x.get("media_type") in ("movie", "tv")
            ]

        # Attempt 1 & 2 — type-specific, with then without year
        for with_year in (True, False):
            results = await _typed(title, with_year)
            if results:
                results.sort(key=_score, reverse=True)
                return _extract(results[0], title_key)

        # Attempt 3 — TMDB multi-search (handles cross-type and fuzzy spelling)
        results = await _multi(title)
        if results:
            results.sort(key=_score, reverse=True)
            best  = results[0]
            tkey  = "name" if best.get("media_type") == "tv" else "title"
            return _extract(best, tkey)

        # Attempt 4 — progressively shorten the title (handles over-long
        # filenames like "Demon.Slayer.Kimetsu.no.Yaiba.Infinity.Castle.2025")
        words = title.split()
        for drop in range(1, min(5, len(words) - 1)):   # keep at least 2 words
            shorter  = " ".join(words[:-drop])
            results  = await _multi(shorter)
            if results:
                results.sort(key=_score, reverse=True)
                best  = results[0]
                tkey  = "name" if best.get("media_type") == "tv" else "title"
                logger.info(
                    "fuzzy_resolve: matched '%s' via shortened query '%s'",
                    best.get(tkey), shorter,
                )
                return _extract(best, tkey)

        raise ValueError(f"No TMDB results for '{title}'")

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    @staticmethod
    def _parse_query(query: str) -> tuple[str, Optional[int], Optional[int]]:
        """Strip season/episode tags from *query* and return them separately."""
        season = episode = None

        # S01E03 style
        m = _SE_FULL.search(query)
        if m:
            season, episode = int(m.group(1)), int(m.group(2))
            query = query[:m.start()].strip()
            return query, season, episode

        # "Season 2" / "season 2"
        m = _SEASON_WORD.search(query)
        if m:
            season = int(m.group(1))
            query = query[:m.start()].strip()

        # S01 style
        if season is None:
            m = _S_ONLY.search(query)
            if m:
                season = int(m.group(1))
                query = query[:m.start()].strip()

        # "Episode 3"
        m = _EPISODE_WORD.search(query)
        if m:
            episode = int(m.group(1))
            query = query[:m.start()].strip()

        # Strip trailing year from clean query
        query = _YEAR_TRAIL.sub("", query).strip()
        return query, season, episode

    async def _search_multi(
        self, query: str, season: Optional[int], episode: Optional[int]
    ) -> MediaInfo:
        """Use TMDB's /search/multi and pick movie vs TV based on result confidence."""
        data = await self._get("search/multi", params={"query": query, "include_adult": False})
        results = data.get("results", [])

        # Filter to movie/tv, sort by popularity
        candidates = [x for x in results if x.get("media_type") in ("movie", "tv")]
        if not candidates:
            raise ValueError(f"No TMDB results for '{query}'")

        candidates.sort(key=lambda x: x.get("popularity", 0), reverse=True)
        best = candidates[0]
        media_type = best["media_type"]

        if media_type == "movie":
            return await self._build_movie_info(best)
        else:
            return await self._build_tv_info(best, season, episode)

    async def _search_tv(
        self, query: str, season: Optional[int], episode: Optional[int]
    ) -> MediaInfo:
        data = await self._get("search/tv", params={"query": query})
        results = data.get("results", [])
        if not results:
            raise ValueError(f"No TV show found for '{query}'")
        results.sort(key=lambda x: x.get("popularity", 0), reverse=True)
        return await self._build_tv_info(results[0], season, episode)

    async def _lookup_imdb(
        self, imdb_id: str, season: Optional[int], episode: Optional[int]
    ) -> MediaInfo:
        data = await self._get("find/" + imdb_id, params={"external_source": "imdb_id"})
        if data.get("movie_results"):
            return await self._build_movie_info(data["movie_results"][0])
        if data.get("tv_results"):
            return await self._build_tv_info(data["tv_results"][0], season, episode)
        raise ValueError(f"IMDb ID {imdb_id} not found on TMDB")

    async def _build_movie_info(self, tmdb_result: dict) -> MediaInfo:
        tmdb_id = tmdb_result["id"]

        # Fetch external IDs for IMDb
        ext = await self._get(f"movie/{tmdb_id}/external_ids")
        imdb_id = ext.get("imdb_id", "")

        # Fetch full movie details for genres
        details = await self._get(f"movie/{tmdb_id}")

        title = details.get("title") or tmdb_result.get("title", "Unknown")
        year_str = (details.get("release_date") or "")[:4]
        year = int(year_str) if year_str.isdigit() else None
        genres = [g["id"] for g in details.get("genres", [])]
        is_anime = 16 in genres and details.get("original_language") == "ja"

        return MediaInfo(
            title=title,
            year=year,
            imdb_id=imdb_id,
            tmdb_id=tmdb_id,
            type="anime" if is_anime else "movie",
            is_anime=is_anime,
            overview=details.get("overview", ""),
            poster_path=details.get("poster_path") or tmdb_result.get("poster_path"),
        )

    async def _build_tv_info(
        self,
        tmdb_result: dict,
        season: Optional[int],
        episode: Optional[int],
    ) -> MediaInfo:
        tmdb_id = tmdb_result["id"]

        ext = await self._get(f"tv/{tmdb_id}/external_ids")
        imdb_id = ext.get("imdb_id", "")

        details = await self._get(f"tv/{tmdb_id}")

        title = details.get("name") or tmdb_result.get("name", "Unknown")
        year_str = (details.get("first_air_date") or "")[:4]
        year = int(year_str) if year_str.isdigit() else None
        genres = [g["id"] for g in details.get("genres", [])]
        origin = details.get("origin_country", [])
        is_anime = (
            16 in genres
            and (details.get("original_language") == "ja" or "JP" in origin)
        ) or ANIME_KEYWORDS.search(title) is not None

        total_seasons = details.get("number_of_seasons")
        media_type = "anime" if is_anime else "tv"

        # If a season was requested, fetch its episode list
        ep_count = 0
        ep_titles: dict[int, str] = {}
        if season is not None:
            try:
                s_data = await self._get(f"tv/{tmdb_id}/season/{season}")
                episodes = s_data.get("episodes", [])
                ep_count = len(episodes)
                ep_titles = {e["episode_number"]: e.get("name", "") for e in episodes}
            except Exception as exc:
                logger.warning("Failed to fetch season %d for tmdb=%d: %s", season, tmdb_id, exc)

        return MediaInfo(
            title=title,
            year=year,
            imdb_id=imdb_id,
            tmdb_id=tmdb_id,
            type=media_type,
            season=season,
            episode=episode,
            total_seasons=total_seasons,
            episodes_in_season=ep_count or None,
            episode_titles=ep_titles,
            is_anime=is_anime,
            overview=details.get("overview", ""),
            poster_path=details.get("poster_path") or tmdb_result.get("poster_path"),
        )

File: app/MediaDownloader/MediaDownloader.Server/Clients/TmdbClient.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace MediaDownloader.Server.Clients;

/// <summary>
/// TMDB API client — parses queries and resolves to structured media info.
/// Mirrors server/clients/tmdb_client.py.
/// </summary>
public class TmdbClient : BaseApiClient
{
    private const string TmdbBase = "https://api.themoviedb.org/3/";
    private readonly string _apiKey;

    private static readonly Regex AnimeKeywords = new(
        @"\b(anime|manga|ova|ona|shounen|shonen|shoujo|seinen|isekai|mecha)\b",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex SeFull = new(@"[Ss](\d{1,2})[Ee](\d{1,3})", RegexOptions.Compiled);
    private static readonly Regex SOnly = new(@"[Ss](\d{1,2})\b", RegexOptions.Compiled);
    private static readonly Regex SeasonWord = new(@"\bseason\s+(\d{1,2})\b", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex EpisodeWord = new(@"\bepisode\s+(\d{1,3})\b", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex ImdbUrl = new(@"imdb\.com/title/(tt\d+)", RegexOptions.Compiled);
    private static readonly Regex YearTrail = new(@"\s*\(?(19|20)\d{2}\)?$", RegexOptions.Compiled);

    public TmdbClient(string apiKey, ILogger? logger = null)
        : base(TmdbBase, timeoutSeconds: 15, maxRetries: 3, backoffBase: 1.0, logger: logger)
    {
        _apiKey = apiKey;
    }

    private async Task<JsonElement> TmdbGet(string path, Dictionary<string, string>? extraParams = null, CancellationToken ct = default)
    {
        path = path.TrimStart('/');
        var sep = path.Contains('?') ? '&' : '?';
        var url = $"{path}{sep}api_key={_apiKey}";
        if (extraParams != null)
        {
            foreach (var (k, v) in extraParams)
                url += $"&{Uri.EscapeDataString(k)}={Uri.EscapeDataString(v)}";
        }
        return await GetJsonAsync(url, ct);
    }

    // ------------------------------------------------------------------
    // Public API
    // ------------------------------------------------------------------

    public async Task<MediaInfo> SearchAsync(string rawQuery, CancellationToken ct = default)
    {
        var (query, season, episode) = ParseQuery(rawQuery.Trim());

        var imdbMatch = ImdbUrl.Match(rawQuery);
        if (imdbMatch.Success)
            return await LookupImdbAsync(imdbMatch.Groups[1].Value, season, episode, ct);

        if (season.HasValue || episode.HasValue)
            return await SearchTvAsync(query, season, episode, ct);

        return await SearchMultiAsync(query, season, episode, ct);
    }

    public async Task<int> GetEpisodeCountAsync(int tmdbId, int season, CancellationToken ct = default)
    {
        try
        {
            var data = await TmdbGet($"tv/{tmdbId}/season/{season}", ct: ct);
            return data.TryGetProperty("episodes", out var eps) ? eps.GetArrayLength() : 0;
        }
        catch { return 0; }
    }

    public async Task<string> GetEpisodeTitleAsync(int tmdbId, int season, int episode, CancellationToken ct = default)
    {
        try
        {
            var data = await TmdbGet($"tv/{tmdbId}/season/{season}/episode/{episode}", ct: ct);
            return data.TryGetProperty("name", out var name) ? name.GetString() ?? "" : "";
        }
        catch { return ""; }
    }

    public async Task<(string title, int? year, string? posterPath)> FuzzyResolveAsync(
        string title, string mediaType = "movie", int? year = null, CancellationToken ct = default)
    {
        bool isTv = mediaType is "tv" or "anime";
        string endpoint = isTv ? "search/tv" : "search/movie";
        string dateKey = isTv ? "first_air_date_year" : "year";
        string titleKey = isTv ? "name" : "title";

        (string, int?, string?) Extract(JsonElement best, string tkey)
        {
            var ds = GetString(best, "release_date") ?? GetString(best, "first_air_date") ?? "";
            var t = GetString(best, tkey) ?? title;
            var y = ds.Length >= 4 && int.TryParse(ds[..4], out var yy) ? yy : year;
            var pp = GetString(best, "poster_path");
            return (t, y, pp);
        }

        double Score(JsonElement item)
        {
            var t = (GetString(item, "title") ?? GetString(item, "name") ?? "").ToLowerInvariant();
            var ds = GetString(item, "release_date") ?? GetString(item, "first_air_date") ?? "";
            var iy = ds.Length >= 4 && int.TryParse(ds[..4], out var yy) ? yy : 0;
            double s = 0;
            if (t == title.ToLowerInvariant()) s += 100;
            if (t == title.ToLowerInvariant() && (!year.HasValue || iy == year)) s += 200;
            s += item.TryGetProperty("popularity", out var pop) ? pop.GetDouble() : 0;
            return s;
        }

        // Attempt 1 & 2 — type-specific, with then without year
        foreach (var withYear in new[] { true, false })
        {
            var p = new Dictionary<string, string> { ["query"] = title, ["include_adult"] = "false" };
            if (withYear && year.HasValue) p[dateKey] = year.Value.ToString();
            var data = await TmdbGet(endpoint, p, ct);
            if (data.TryGetProperty("results", out var results) && results.GetArrayLength() > 0)
            {
                var best = results.EnumerateArray().OrderByDescending(Score).First();
                return Extract(best, titleKey);
            }
        }

        // Attempt 3 — multi-search
        var multi = await TmdbGet("search/multi",
            new Dictionary<string, string> { ["query"] = title, ["include_adult"] = "false" }, ct);
        if (multi.TryGetProperty("results", out var mr))
        {
            var filtered = mr.EnumerateArray()
                .Where(x => GetString(x, "media_type") is "movie" or "tv")
                .ToList();
            if (filtered.Count > 0)
            {
                var best = filtered.OrderByDescending(Score).First();
                var tkey = GetString(best, "media_type") == "tv" ? "name" : "title";
                return Extract(best, tkey);
            }
        }

        // Attempt 4 — progressively shorten title
        var words = title.Split(' ');
        for (int drop = 1; drop < Math.Min(5, words.Length); drop++)
        {
            var shorter = string.Join(' ', words[..^drop]);
            var shortMulti = await TmdbGet("search/multi",
                new Dictionary<string, string> { ["query"] = shorter, ["include_adult"] = "false" }, ct);
            if (shortMulti.TryGetProperty("results", out var sr))
            {
                var filtered = sr.EnumerateArray()
                    .Where(x => GetString(x, "media_type") is "movie" or "tv")
                    .ToList();
                if (filtered.Count > 0)
                {
                    var best = filtered.OrderByDescending(Score).First();
                    var tkey = GetString(best, "media_type") == "tv" ? "name" : "title";
                    return Extract(best, tkey);
                }
            }
        }

        throw new InvalidOperationException($"No TMDB results for '{title}'");
    }

    // ------------------------------------------------------------------
    // Internal helpers
    // ------------------------------------------------------------------

    internal static (string query, int? season, int? episode) ParseQuery(string query)
    {
        int? season = null, episode = null;

        var m = SeFull.Match(query);
        if (m.Success)
        {
            season = int.Parse(m.Groups[1].Value);
            episode = int.Parse(m.Groups[2].Value);
            return (query[..m.Index].Trim(), season, episode);
        }

        m = SeasonWord.Match(query);
        if (m.Success)
        {
            season = int.Parse(m.Groups[1].Value);
            query = query[..m.Index].Trim();
        }

        if (!season.HasValue)
        {
            m = SOnly.Match(query);
            if (m.Success)
            {
                season = int.Parse(m.Groups[1].Value);
                query = query[..m.Index].Trim();
            }
        }

        m = EpisodeWord.Match(query);
        if (m.Success)
        {
            episode = int.Parse(m.Groups[1].Value);
            query = query[..m.Index].Trim();
        }

        query = YearTrail.Replace(query, "").Trim();
        return (query, season, episode);
    }

    private async Task<MediaInfo> SearchMultiAsync(string query, int? season, int? episode, CancellationToken ct)
    {
        var data = await TmdbGet("search/multi",
            new Dictionary<string, string> { ["query"] = query, ["include_adult"] = "false" }, ct);

        if (!data.TryGetProperty("results", out var results))
            throw new InvalidOperationException($"No TMDB results for '{query}'");

        var candidates = results.EnumerateArray()
            .Where(x => GetString(x, "media_type") is "movie" or "tv")
            .OrderByDescending(x => x.TryGetProperty("popularity", out var p) ? p.GetDouble() : 0)
            .ToList();

        if (candidates.Count == 0)
            throw new InvalidOperationException($"No TMDB results for '{query}'");

        var best = candidates[0];
        return GetString(best, "media_type") == "movie"
            ? await BuildMovieInfoAsync(best, ct)
            : await BuildTvInfoAsync(best, season, episode, ct);
    }

    private async Task<MediaInfo> SearchTvAsync(string query, int? season, int? episode, CancellationToken ct)
    {
        var data = await TmdbGet("search/tv",
            new Dictionary<string, string> { ["query"] = query }, ct);

        if (!data.TryGetProperty("results", out var results) || results.GetArrayLength() == 0)
            throw new InvalidOperationException($"No TV show found for '{query}'");

        var best = results.EnumerateArray()
            .OrderByDescending(x => x.TryGetProperty("popularity", out var p) ? p.GetDouble() : 0)
            .First();

        return await BuildTvInfoAsync(best, season, episode, ct);
    }

    private async Task<MediaInfo> LookupImdbAsync(string imdbId, int? season, int? episode, CancellationToken ct)
    {
        var data = await TmdbGet($"find/{imdbId}",
            new Dictionary<string, string> { ["external_source"] = "imdb_id" }, ct);

        if (data.TryGetProperty("movie_results", out var movies) && movies.GetArrayLength() > 0)
            return await BuildMovieInfoAsync(movies[0], ct);
        if (data.TryGetProperty("tv_results", out var tvs) && tvs.GetArrayLength() > 0)
            return await BuildTvInfoAsync(tvs[0], season, episode, ct);

        throw new InvalidOperationException($"IMDb ID {imdbId} not found on TMDB");
    }

    private async Task<MediaInfo> BuildMovieInfoAsync(JsonElement tmdbResult, CancellationToken ct)
    {
        var tmdbId = tmdbResult.GetProperty("id").GetInt32();
        var ext = await TmdbGet($"movie/{tmdbId}/external_ids", ct: ct);
        var details = await TmdbGet($"movie/{tmdbId}", ct: ct);

        var title = GetString(details, "title") ?? GetString(tmdbResult, "title") ?? "Unknown";
        var yearStr = (GetString(details, "release_date") ?? "")[..Math.Min(4, (GetString(details, "release_date") ?? "").Length)];
        var year = yearStr.Length == 4 && int.TryParse(yearStr, out var y) ? y : (int?)null;

        var genres = details.TryGetProperty("genres", out var g)
            ? g.EnumerateArray().Select(x => x.GetProperty("id").GetInt32()).ToHashSet()
            : new HashSet<int>();
        var isAnime = genres.Contains(16) && GetString(details, "original_language") == "ja";

        return new MediaInfo
        {
            Title = title, Year = year,
            ImdbId = GetString(ext, "imdb_id") ?? "",
            TmdbId = tmdbId,
            Type = isAnime ? "anime" : "movie",
            IsAnime = isAnime,
            Overview = GetString(details, "overview") ?? "",
            PosterPath = GetString(details, "poster_path") ?? GetString(tmdbResult, "poster_path"),
        };
    }

    private async Task<MediaInfo> BuildTvInfoAsync(JsonElement tmdbResult, int? season, int? episode, CancellationToken ct)
    {
        var tmdbId = tmdbResult.GetProperty("id").GetInt32();
        var ext = await TmdbGet($"tv/{tmdbId}/external_ids", ct: ct);
        var details = await TmdbGet($"tv/{tmdbId}", ct: ct);

        var title = GetString(details, "name") ?? GetString(tmdbResult, "name") ?? "Unknown";
        var yearStr = (GetString(details, "first_air_date") ?? "")[..Math.Min(4, (GetString(details, "first_air_date") ?? "").Length)];
        var year = yearStr.Length == 4 && int.TryParse(yearStr, out var y) ? y : (int?)null;

        var genres = details.TryGetProperty("genres", out var g)
            ? g.EnumerateArray().Select(x => x.GetProperty("id").GetInt32()).ToHashSet()
            : new HashSet<int>();
        var origin = details.TryGetProperty("origin_country", out var oc)
            ? oc.EnumerateArray().Select(x => x.GetString() ?? "").ToHashSet()
            : new HashSet<string>();
        var isAnime = (genres.Contains(16) &&
            (GetString(details, "original_language") == "ja" || origin.Contains("JP")))
            || AnimeKeywords.IsMatch(title);

        var totalSeasons = details.TryGetProperty("number_of_seasons", out var ns) ? ns.GetInt32() : (int?)null;

        int? epCount = null;
        var epTitles = new Dictionary<int, string>();
        if (season.HasValue)
        {
            try
            {
                var sData = await TmdbGet($"tv/{tmdbId}/season/{season}", ct: ct);
                if (sData.TryGetProperty("episodes", out var eps))
                {
                    epCount = eps.GetArrayLength();
                    foreach (var ep in eps.EnumerateArray())
                    {
                        var epNum = ep.GetProperty("episode_number").GetInt32();
                        var epName = GetString(ep, "name") ?? "";
                        epTitles[epNum] = epName;
                    }
                }
            }
            catch { /* ignore */ }
        }

        return new MediaInfo
        {
            Title = title, Year = year,
            ImdbId = GetString(ext, "imdb_id") ?? "",
            TmdbId = tmdbId,
            Type = isAnime ? "anime" : "tv",
            Season = season, Episode = episode,
            TotalSeasons = totalSeasons,
            EpisodesInSeason = epCount,
            EpisodeTitles = epTitles,
            IsAnime = isAnime,
            Overview = GetString(details, "overview") ?? "",
            PosterPath = GetString(details, "poster_path") ?? GetString(tmdbResult, "poster_path"),
        };
    }

    private static string? GetString(JsonElement el, string prop)
        => el.TryGetProperty(prop, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;
}
